default/AI/UniverseUtilsAI.py

```diff
@@ -0,0 +1,164 @@
+    # a Sentinel or somesuch might throw it off, That is an area for future improvement
+    return fo.getEmpire().supplyProjections(-1-get_potential_supply_distance(), False)
+
+
+@cache_by_turn
+def _all_systems_by_supply_tier():
+    """Get planets by distance to supply lines.
+
+    :return: indexed by the negative value of distance to supplied systems, contains list of system ids.
+    :rtype: dict
+    """
+    systems_by_supply_tier = {}
+    for sys_id, supply_val in get_system_supply().items():
+        systems_by_supply_tier.setdefault(min(0, supply_val), []).append(sys_id)
+    print
+    print "New Supply Calc:"
+    print "Known Systems:", fo.getUniverse().systemIDs
+    print "Base Supply:", dict_from_map(fo.getEmpire().systemSupplyRanges)
+    print "New Supply connected systems: ",
+    print ppstring(PlanetUtilsAI.sys_name_ids(systems_by_supply_tier.get(0, [])))
+    print "New First Ring of annexable systems: ",
+    print ppstring(PlanetUtilsAI.sys_name_ids(systems_by_supply_tier.get(-1, [])))
+    print "New Second Ring of annexable systems: ",
+    print ppstring(PlanetUtilsAI.sys_name_ids(systems_by_supply_tier.get(-2, [])))
+    print "New Third Ring of annexable systems: ",
+    print ppstring(PlanetUtilsAI.sys_name_ids(systems_by_supply_tier.get(-3, [])))
+    return systems_by_supply_tier
+
+
+def get_systems_by_supply_tier(supply_tier=None):
+    """Get all systems by distance to supply lines.
+
+    If no supply_tier is specified, return a dict of lists for each distance.
+
+    :param supply_tier: Distance to empire supply lines.
+    :type supply_tier: int
+    :return: List of system IDs.
+    :rtype: list or dict
+    """
+    if supply_tier is None:
+        return _all_systems_by_supply_tier()
+    return _all_systems_by_supply_tier().get(supply_tier, [])
+
+
+@cache_by_turn
+def get_potential_supply_distance():
+    """Find the maximum possible supply distance for this empire.
+
+    Considers techs, planet size, species.
+
+    :rtype: int
+    """
+    potential_supply_distance = get_supply_tech_range()
+    potential_supply_distance += max(AIDependencies.supply_by_size.values())
+    potential_supply_distance += max(AIDependencies.species_supply_range_modifier.values())
+    potential_supply_distance += 1  # World tree special
+    # TODO: +1 to consider capturing planets with Elevators
+    # if foAI.foAIstate.aggression >= fo.aggression.aggressive:
+    # potential_supply_distance += 1
+    return potential_supply_distance
+
+
+@cache_by_turn
+def get_all_annexable_system_ids():
+    """Get all systems considered annexable.
+
+    :return: of system IDs
+    :rtype: set
+    """
+    # TODO: distinguish colony-annexable systems and outpost-annexable systems
+    # TODO: Make sure, that update annexable systm got called before
+    annexable_system_ids = set()
+    systems_by_supply_tier = get_systems_by_supply_tier()
+    for jumps in range(0, -1-get_potential_supply_distance(), -1):
```

Comment:
You iterate in backwards, but it does not matter, you don't store order of items. You can simplify your `range`